Issue #11495 - Add UriCompliance rules that follow the HTTP / URI / Servlet specs for illegal & suspicious characters

joakime opened the pull request:
Adds 2 more UriCompliance.Violations

* `ILLEGAL_PATH_CHARACTERS` for both the URI and HTTP spec rules about illegal (not encoded) characters
* `SUSPICIOUS_PATH_CHARACTERS` for Servlet spec rules on suspicious (encoded or not) characters

Fixes: #11495

joakime committed:
Issue #11495 - Add UriCompliance rules that follow the HTTP / URI / Servlet specs for illegal & suspicious characters

joakime commented:
See https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0.html#uri-path-canonicalization for list of Servlet rules. (point 10)

joakime committed:
more illegalPathCharacterData test cases

joakime committed:
Fix comment

gregw commented on jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java:
```diff
@@ -641,6 +642,91 @@ public void testPathQuery(String input, String canonicalPath, String decodedPath
+    {
+        return Stream.of(
+            // backslash
+            Arguments.of("/a%5Cb"),
+            Arguments.of("/foo/bar/zed/%5c"),
+            Arguments.of("/foo/bar/zed/%5C"),
```
The spec says: 
> The "\\" character encoded or not. (e.g. /path\info)
So can we add some tests for non encoded \ ?

joakime replied:
Done.   pushed a fix to make the semantic more in line with what the Servlet spec says.

joakime committed:
Correcting SUSPICIOUS_PATH_CHARACTERS semantic (encoded and decoded)

joakime committed:
Correct param handling

gregw committed:
Optimizations

Check for illegal and suspicious characters as we are parsing the path.
Only look for ambiguous paths if we know there are dots or encodings.

gregw committed:
preserve parameter

gregw committed:
fixed tests

gregw committed:
Fix tests

gregw committed:
Fix tests

joakime committed:
Updating tests

joakime committed:
Merge remote-tracking branch 'origin/fix/12.0.x/uricompliance-reject-backslash' into fix/12.0.x/uricompliance-reject-backslash

# Conflicts:
#	jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DefaultServletTest.java

gregw commented:
@joakime status of this?

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/12.0.x/uricompliance-reject-backslash

joakime commented:
@gregw this PR is green and ready for review

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -895,17 +970,18 @@ public Mutable path(String path)
             _canonicalPath = null;
+            String param = _param;
+            _param = null;
+            parse(State.PATH, path);
```
I'm cautious about this, as this now allows for a query and fragment to be passed in via this method.  Can this be done without a call to parse?  perhaps we should check for ? and # characters before parsing... or does the isPathValid already do that?
Note we should javadoc if the path is raw or encoded.

joakime replied:
There's a fair number of places that uses this `HttpURI.Mutable.path(String)` API.
Interestingly, we actually need to parse the path in some of those places to satisfy the Servlet 6.0 "suspicious characters" rule. (eg: `sendRedirect`, `getRequestDispatcher`, etc)
I'm not sure what `isPathValid` you are mentioning here.
It would be great if we had a `parsePath` that only did the path, and if it encountered something outside of the scope of PATH (eg: query / fragment / etc) that would also be an error.

joakime replied:
> Note we should javadoc if the path is raw or encoded.

As far as I'm concerned, this behavior didn't change for `HttpURI.Mutable.path(String)`.
But since it wasn't originally documented, I'm not sure the intent of this method.

joakime replied:
@gregw bump

gregw replied:
@joakime I've pushed a fix from my concern.... if you don't like it, we can revert and I'll OK this anyway.

gregw reviewed (changes requested):
Almost... but one question needs to be answered... plus a little javadoc

gregw committed:
Safer parse by not allowing beyond a final state

joakime committed:
Revert "Safer parse by not allowing beyond a final state"

This reverts commit fed10f79aacecfcb4118bcbe39fcb21d98666f25.
